File: sdk-typescript/packages/studio/src/studio.ts
/**
 * WAVE SDK - Studio API
 *
 * Multi-camera broadcast production system for creating, managing, and
 * controlling live productions with sources, scenes, graphics, and audio mixing.
 *
 * NOTE: This is a client SDK. All authorization checks are performed server-side.
 * The API will return 403 Forbidden if the user lacks required permissions.
 */

import type {
  WaveClient,
  PaginatedResponse,
} from '@wave-av/core';

import type {
  AudioMixChannel,
  CreateProductionRequest,
  Graphic,
  LayoutType,
  ListProductionsParams,
  Production,
  Scene,
  SceneSource,
  Source,
  SourceType,
  TransitionConfig,
  UpdateProductionRequest,
} from './studio-types';
export type { ProductionStatus, SourceType, TransitionType, LayoutType, TransitionConfig, SceneSource, Scene, Source, Production, Graphic, AudioMixChannel, CreateProductionRequest, UpdateProductionRequest, ListProductionsParams } from './studio-types';

export class StudioAPI {
  private readonly client: WaveClient;
  private readonly basePath = "/v1/productions";

  constructor(client: WaveClient) {
    this.client = client;
  }


  /**
   * Create a new production
   *
   * Requires: productions:create permission
   */
  async create(request: CreateProductionRequest): Promise<Production> {
    return this.client.post<Production>(this.basePath, request);
  }

  /**
   * Get a production by ID
   *
   * Requires: productions:read permission
   */
  async get(productionId: string): Promise<Production> {
    return this.client.get<Production>(`${this.basePath}/${productionId}`);
  }

  /**
   * Update a production
   *
   * Requires: productions:update permission
   */
  async update(productionId: string, request: UpdateProductionRequest): Promise<Production> {
    return this.client.patch<Production>(`${this.basePath}/${productionId}`, request);
  }

  /**
   * Remove a production
   *
   * Requires: productions:remove permission (server-side RBAC enforced)
   */
  async remove(productionId: string): Promise<void> {
    await this.client.delete(`${this.basePath}/${productionId}`);
  }

  /**
   * List productions with optional filters
   *
   * Requires: productions:read permission
   */
  async list(params?: ListProductionsParams): Promise<PaginatedResponse<Production>> {
    const queryParams: Record<string, string | number | boolean | undefined> = {
      limit: params?.limit,
      offset: params?.offset,
      cursor: params?.cursor,
      status: params?.status,
      created_after: params?.created_after,
      created_before: params?.created_before,
      order_by: params?.order_by,
      order: params?.order,
    };

    return this.client.get<PaginatedResponse<Production>>(this.basePath, {
      params: queryParams,
    });
  }


  /**
   * Start a production (go live)
   *
   * Transitions the production from 'idle' or 'rehearsal' to 'live'.
   *
   * Requires: productions:control permission
   */
  async start(productionId: string): Promise<Production> {
    return this.client.post<Production>(`${this.basePath}/${productionId}/start`);
  }

  /**
   * Stop a production (end broadcast)
   *
   * Transitions the production to 'ending' and then 'ended'.
   *
   * Requires: productions:control permission
   */
  async stop(productionId: string): Promise<Production> {
    return this.client.post<Production>(`${this.basePath}/${productionId}/stop`);
  }

  /**
   * Start a rehearsal session
   *
   * Allows testing sources, scenes, and transitions without going live.
   * Transitions the production from 'idle' to 'rehearsal'.
   *
   * Requires: productions:control permission
   */
  async startRehearsal(productionId: string): Promise<Production> {
    return this.client.post<Production>(`${this.basePath}/${productionId}/rehearsal`);
  }


  /**
   * Add an input source to a production
   *
   * Requires: productions:sources:create permission
   */
  async addSource(
    productionId: string,
    source: {
      name: string;
      type: SourceType;
      url?: string;
      ptz_enabled?: boolean;
    },
  ): Promise<Source> {
    return this.client.post<Source>(`${this.basePath}/${productionId}/sources`, source);
  }

  /**
   * Remove a source from a production
   *
   * Requires: productions:sources:remove permission
   */
  async removeSource(productionId: string, sourceId: string): Promise<void> {
    await this.client.delete(`${this.basePath}/${productionId}/sources/${sourceId}`);
  }

  /**
   * List all sources for a production
   *
   * Requires: productions:sources:read permission
   */
  async listSources(productionId: string): Promise<Source[]> {
    return this.client.get<Source[]>(`${this.basePath}/${productionId}/sources`);
  }

  /**
   * Get a specific source by ID
   *
   * Requires: productions:sources:read permission
   */
  async getSource(productionId: string, sourceId: string): Promise<Source> {
    return this.client.get<Source>(`${this.basePath}/${productionId}/sources/${sourceId}`);
  }


  /**
   * Create a new scene in a production
   *
   * Requires: productions:scenes:create permission
   */
  async createScene(
    productionId: string,
    scene: {
      name: string;
      layout: LayoutType;
      sources?: SceneSource[];
      transition_in?: TransitionConfig;
    },
  ): Promise<Scene> {
    return this.client.post<Scene>(`${this.basePath}/${productionId}/scenes`, scene);
  }

  /**
   * Update an existing scene
   *
   * Requires: productions:scenes:update permission
   */
  async updateScene(
    productionId: string,
    sceneId: string,
    updates: Partial<{
      name: string;
      layout: LayoutType;
      sources: SceneSource[];
      transition_in: TransitionConfig;
      sort_order: number;
    }>,
  ): Promise<Scene> {
    return this.client.patch<Scene>(`${this.basePath}/${productionId}/scenes/${sceneId}`, updates);
  }

  /**
   * Remove a scene from a production
   *
   * Requires: productions:scenes:remove permission
   */
  async removeScene(productionId: string, sceneId: string): Promise<void> {
    await this.client.delete(`${this.basePath}/${productionId}/scenes/${sceneId}`);
  }

  /**
   * List all scenes for a production
   *
   * Requires: productions:scenes:read permission
   */
  async listScenes(productionId: string): Promise<Scene[]> {
    return this.client.get<Scene[]>(`${this.basePath}/${productionId}/scenes`);
  }

  /**
   * Activate a scene with an optional transition
   *
   * Sets the scene as the active scene for the production output.
   *
   * Requires: productions:scenes:control permission
   */
  async activateScene(
    productionId: string,
    sceneId: string,
    transition?: TransitionConfig,
  ): Promise<Scene> {
    return this.client.post<Scene>(
      `${this.basePath}/${productionId}/scenes/${sceneId}/activate`,
      transition ? { transition } : undefined,
    );
  }


  /**
   * Set the program (live) source with an optional transition
   *
   * Switches the currently live output to the specified source.
   *
   * Requires: productions:control permission
   */
  async setProgram(
    productionId: string,
    sourceId: string,
    transition?: TransitionConfig,
  ): Promise<void> {
    await this.client.post(`${this.basePath}/${productionId}/program`, {
      source_id: sourceId,
      transition,
    });
  }

  /**
   * Set the preview source
   *
   * Loads a source into the preview output for inspection before going live.
   *
   * Requires: productions:control permission
   */
  async setPreview(productionId: string, sourceId: string): Promise<void> {
    await this.client.post(`${this.basePath}/${productionId}/preview`, { source_id: sourceId });
  }

  /**
   * Execute a transition between preview and program
   *
   * Swaps the current preview source into program using the specified transition.
   *
   * Requires: productions:control permission
   */
  async transition(productionId: string, config: TransitionConfig): Promise<void> {
    await this.client.post(`${this.basePath}/${productionId}/transition`, config);
  }


  /**
   * Add a graphic overlay to a production
   *
   * Requires: productions:graphics:create permission
   */
  async addGraphic(
    productionId: string,
    graphic: {
      name: string;
      type: Graphic["type"];
      content: Record<string, unknown>;
      position?: Graphic["position"];
      layer?: number;
    },
  ): Promise<Graphic> {
    return this.client.post<Graphic>(`${this.basePath}/${productionId}/graphics`, graphic);
  }

  /**
   * Update an existing graphic
   *
   * Requires: productions:graphics:update permission
   */
  async updateGraphic(
    productionId: string,
    graphicId: string,
    updates: Partial<{
      name: string;
      type: Graphic["type"];
      content: Record<string, unknown>;
      position: Graphic["position"];
      layer: number;
    }>,
  ): Promise<Graphic> {
    return this.client.patch<Graphic>(
      `${this.basePath}/${productionId}/graphics/${graphicId}`,
      updates,
    );
  }

  /**
   * Remove a graphic from a production
   *
   * Requires: productions:graphics:remove permission
   */
  async removeGraphic(productionId: string, graphicId: string): Promise<void> {
    await this.client.delete(`${this.basePath}/${productionId}/graphics/${graphicId}`);
  }

  /**
   * Show a graphic on the production output
   *
   * Makes the graphic visible on the live output.
   *
   * Requires: productions:graphics:control permission
   */
  async showGraphic(productionId: string, graphicId: string): Promise<void> {
    await this.client.post(`${this.basePath}/${productionId}/graphics/${graphicId}/show`);
  }

  /**
   * Hide a graphic from the production output
   *
   * Removes the graphic from the live output without deleting it.
   *
   * Requires: productions:graphics:control permission
   */
  async hideGraphic(productionId: string, graphicId: string): Promise<void> {
    await this.client.post(`${this.basePath}/${productionId}/graphics/${graphicId}/hide`);
  }


  /**
   * Get the current audio mix for a production
   *
   * Returns volume, mute, solo, pan, and processing settings for all channels.
   *
   * Requires: productions:audio:read permission
   */
  async getAudioMix(productionId: string): Promise<AudioMixChannel[]> {
    return this.client.get<AudioMixChannel[]>(`${this.basePath}/${productionId}/audio-mix`);
  }

  /**
   * Set the audio mix for a production
   *
   * Updates volume, mute, solo, pan, and processing settings for channels.
   *
   * Requires: productions:audio:control permission
   */
  async setAudioMix(productionId: string, channels: AudioMixChannel[]): Promise<AudioMixChannel[]> {
    return this.client.put<AudioMixChannel[]>(`${this.basePath}/${productionId}/audio-mix`, {
      channels,
    });
  }
}

/**
 * Create a Studio API instance
 */
export function createStudioAPI(client: WaveClient): StudioAPI {
  return new StudioAPI(client);
}

File: sdk-python/wave_sdk/studio.py
"""WAVE SDK - Studio API. Multi-camera broadcast production."""
from __future__ import annotations
from typing import Any
from pydantic import BaseModel
from wave.client import WaveClient

class Production(BaseModel):
    id: str; organization_id: str; title: str; description: str | None = None; status: str; program_source_id: str | None = None; preview_source_id: str | None = None; active_scene_id: str | None = None; recording_enabled: bool = False; streaming_enabled: bool = False; viewer_count: int = 0; started_at: str | None = None; ended_at: str | None = None; created_at: str; updated_at: str

class Source(BaseModel):
    id: str; production_id: str; name: str; type: str; url: str | None = None; status: str; tally: str; audio_level: float | None = None; volume: float = 1.0; muted: bool = False; ptz_enabled: bool = False

class Scene(BaseModel):
    id: str; production_id: str; name: str; layout: str; sources: list[dict[str, Any]] | None = None; is_active: bool = False; sort_order: int = 0

class Graphic(BaseModel):
    id: str; production_id: str; name: str; type: str; content: dict[str, Any] | None = None; position: dict[str, Any] | None = None; visible: bool = False; layer: int = 0

class StudioAPI:
    def __init__(self, client: WaveClient): self._client = client; self._base = "/v1/productions"
    def create(self, title: str, **kwargs: Any) -> Production: return Production(**self._client.post(self._base, json={"title": title, **kwargs}))
    def get(self, production_id: str) -> Production: return Production(**self._client.get(f"{self._base}/{production_id}"))
    def update(self, production_id: str, **kwargs: Any) -> Production: return Production(**self._client.patch(f"{self._base}/{production_id}", json=kwargs))
    def remove(self, production_id: str) -> None: self._client.delete(f"{self._base}/{production_id}")
    def list(self, **params: Any) -> dict: return self._client.get(self._base, params={k: v for k, v in params.items() if v is not None})
    def start(self, production_id: str) -> Production: return Production(**self._client.post(f"{self._base}/{production_id}/start"))
    def stop(self, production_id: str) -> Production: return Production(**self._client.post(f"{self._base}/{production_id}/stop"))
    def start_rehearsal(self, production_id: str) -> Production: return Production(**self._client.post(f"{self._base}/{production_id}/rehearsal"))
    def add_source(self, production_id: str, name: str, type: str, url: str | None = None, **kwargs: Any) -> Source: return Source(**self._client.post(f"{self._base}/{production_id}/sources", json={"name": name, "type": type, "url": url, **kwargs}))
    def remove_source(self, production_id: str, source_id: str) -> None: self._client.delete(f"{self._base}/{production_id}/sources/{source_id}")
    def list_sources(self, production_id: str) -> list[Source]: return [Source(**s) for s in self._client.get(f"{self._base}/{production_id}/sources")]
    def create_scene(self, production_id: str, name: str, layout: str = "fullscreen", **kwargs: Any) -> Scene: return Scene(**self._client.post(f"{self._base}/{production_id}/scenes", json={"name": name, "layout": layout, **kwargs}))
    def activate_scene(self, production_id: str, scene_id: str, **kwargs: Any) -> Scene: return Scene(**self._client.post(f"{self._base}/{production_id}/scenes/{scene_id}/activate", json=kwargs if kwargs else None))
    def list_scenes(self, production_id: str) -> list[Scene]: return [Scene(**s) for s in self._client.get(f"{self._base}/{production_id}/scenes")]
    def set_program(self, production_id: str, source_id: str, **kwargs: Any) -> None: self._client.post(f"{self._base}/{production_id}/program", json={"source_id": source_id, **kwargs})
    def set_preview(self, production_id: str, source_id: str) -> None: self._client.post(f"{self._base}/{production_id}/preview", json={"source_id": source_id})
    def transition(self, production_id: str, type: str = "cut", duration_ms: int = 0) -> None: self._client.post(f"{self._base}/{production_id}/transition", json={"type": type, "duration_ms": duration_ms})
    def add_graphic(self, production_id: str, name: str, type: str, **kwargs: Any) -> Graphic: return Graphic(**self._client.post(f"{self._base}/{production_id}/graphics", json={"name": name, "type": type, **kwargs}))
    def show_graphic(self, production_id: str, graphic_id: str) -> None: self._client.post(f"{self._base}/{production_id}/graphics/{graphic_id}/show")
    def hide_graphic(self, production_id: str, graphic_id: str) -> None: self._client.post(f"{self._base}/{production_id}/graphics/{graphic_id}/hide")
    def get_audio_mix(self, production_id: str) -> list[dict]: return self._client.get(f"{self._base}/{production_id}/audio")
    def set_audio_mix(self, production_id: str, channels: list[dict]) -> list[dict]: return self._client.put(f"{self._base}/{production_id}/audio", json={"channels": channels})
